-- Replace the following bullets with your information --

*  DSC 530 
*  Week 3 & 4
*  Rethish Plappuzha Sreedharan Nair

In [46]:
# importing the required packages
import random

# Hands-on data analysis with Pandas

# Chapter 1, Exercise 4

Run the code in the first cell of the exercises.ipynb notebook. It will give you a list of 100 values to work with for the rest of the exercises in this chapter. Be sure to treat these values as a sample of the population.


In [47]:
import random

random.seed(0)
salaries = [round(random.random()*1000000, -3) for i in range(100)]

# Currently, this code generated a list of random values as salaries for this exercise.
# Printing the first few elements of the list
salaries[0:10]

[844000.0,
 758000.0,
 421000.0,
 259000.0,
 511000.0,
 405000.0,
 784000.0,
 303000.0,
 477000.0,
 583000.0]

# Chapter 1, Exercise 5

Using the data from exercise 4, calculate the following statistics without importing anything from the [statistics module](https://docs.python.org/3/library/statistics.html) in the Python standard library, and then confirm your results match up to those that are obtained when using the statistics module (where possible):


Tip from instructor: You can use [Pandas](https://pandas.pydata.org/docs/getting_started/intro_tutorials/06_calculate_statistics.html) or [Numpy](https://numpy.org/doc/stable/reference/routines.statistics.html) for the initial calculation and then check against other Python library/package such as [statistics module](https://docs.python.org/3/library/statistics.html), or [SciPy Statistics](https://docs.scipy.org/doc/scipy/reference/stats.html). Part a of Exercise 5 is completed to illustrate the process for you, you can decide how to go about the other problems.

## Exercise 5(a) Mean - with exmaples

In [48]:
# Run this code to make alias to shorted the typed name
# You can do this as the 1st thing in any notebook or later, but you only need to do this once
import numpy as np
import pandas as pd

In [49]:
# Using numpy.mean
print('Mean: ',np.mean(salaries))

Mean:  585690.0


In [50]:
# Using statistics module
import statistics # no need to re-run this after import
# You have to import this standard library before using its mean function
print('Mean using statistics: ', statistics.mean(salaries))

Mean using statistics:  585690.0


In [51]:
# Using Pandas
# If you choose to use Pandas functions, you have to convert the list to a Series, otherwise it will gives an error
salaries_series = pd.Series(salaries)
# For Pandas function, when you placed a dot after a variable or other functions, it is referred as a "method chaining" appraoch.
print('Mean using pandas: ',salaries_series.mean())

Mean using pandas:  585690.0


In [52]:
# Using SciPy Statistics
from scipy import stats
print('Mean using scipy: ',stats.tmean(salaries))

Mean using scipy:  585690.0


## Exercise 5(b) Median

In [53]:
# using numpy
print('Median ',np.median(salaries))

Median  589000.0


In [54]:
# Using pandas 
pd_sal = pd.Series(salaries)
print('Median: ',pd_sal.median())

Median:  589000.0


In [55]:
# Using statistics module
import statistics

print('Median: ', statistics.median(salaries))

Median:  589000.0


## Exercise 5(c) Mode

Hint from textbook: Check out the [Counter class](https://docs.python.org/3/library/collections.html#collections.Counter) in the collections module of the standard library

In [56]:
# Using Counter 
# Importing the Counter from collections
from collections import Counter
# Creates the Counter object by passing the salaries list
c = Counter(salaries)
# Retreiving the most common element 
print('Mode: ',c.most_common(1))

Mode:  [(477000.0, 3)]


In [57]:
# Using Pandas 
print('Mode: ',pd_sal.mode())

Mode:  0    477000.0
dtype: float64


In [58]:
# Using statistics module
print('Mode: ',statistics.mode(salaries))

Mode:  477000.0


## Exercise 5(d) Sample Variance

In [59]:
# Using numpy
 ## We are passing the "Delta Degrees of Freedom" as 1 to calculate sample variance the denominator should be n-1  
print('Variance: ', np.var(salaries, ddof=1))

Variance:  70664054444.44444


In [60]:
# Using pandas 
print('Variance: ',pd_sal.var())

Variance:  70664054444.44444


In [61]:
# Using statistics module
print('Variance: ',statistics.variance(salaries))

Variance:  70664054444.44444


## Exercise 5(e) Sample Standard Deviation

In [62]:
# Uging numpy
print('Standard Deviation: ',np.std(salaries, ddof=1))

Standard Deviation:  265827.11382484


In [63]:
# Using pandas 

print('Standard Deviation: ',pd_sal.std())

Standard Deviation:  265827.11382484


In [64]:
# Using statistics module
print('Standard Deviation: ',statistics.stdev(salaries))

Standard Deviation:  265827.11382484


## Chapter 1, Exercise 6

Using the data from exercise 4, calculate the following statistics using the functions in the statistics module where appropriate:

## Exercise 6(a) Range

In [65]:
# Hint:there is no function called "Range". However, what does it means to see a range of values?
# Your solution
# the range is a measure of variability that represents the difference between the highest and lowest values in a dataset, 
# indicating the spread or dispersion of the data points. 
# It is calculated by subtracting the minimum value from the maximum value (Range = Maximum - Minimum)
range_val = max(salaries) - min(salaries)
range_val

995000.0

## Exercise 6(b) Coefficient of Variation

In [66]:
# Hint: Coefficient of Variation (CV) is defined as the ratio of standard deviation to the mean in Hands-on Data Analysis with Pandas.
# Your solution
co_var = statistics.stdev(salaries) / statistics.mean(salaries)
co_var

0.45386998894439035

## Exercise 6(c) Interquartile Range

In [67]:
# Hint: Interquartile range (IQR) is presented as the difference of the 75% and 25% quantile in Chapter 1 of Practical Statistics for Data Scientisits Textbook
# OR you can find the values in each quartile and then substract Quantile 3 with Quantile 1 as in Chapter 1 of Hands-on Data Analysis with Pandas
# Your solution

# Getting the quartiles
quartiles = statistics.quantiles(salaries,n=4)
# Getting the Q1 and Q3 from the above
q1 = quartiles[0]
q3 = quartiles[2]
# Computing the IQR by subtracting the Q3 and Q1
iqr = q3 - q1
iqr


421750.0

## Exercise 6(d) Quartile coefficient of dispersion

In [68]:
# Once you determined the 1st and 3rd quantile values from 6(c), you can use it for this calculation
# Refer to the Quartile coefficient of dispersion formula in Chapter 1 of Hands-on Data Analysis with Pandas
# Your solution

# The Quartile Coefficient of Dispersion (QCD) is a robust statistical measure of data spread that uses the 
# first and third quartiles (Q1 and Q3). 
# It is calculated as the ratio of half the interquartile range to the sum of the quartiles, specifically 
# (Q3 - Q1) / (Q3 + Q1). 

q_co_disp = (quartiles[2] - quartiles[0])/(quartiles[2]+quartiles[0])

q_co_disp

0.34491923941934166

## Chapter 1, Exercise 7

Scale the data created in exercise 4 using the following strategies:

## Exercise 7(a) Min-max Scaling (Normalizing)

In [69]:
# Refer to the Scaling data section of your reading in Chapter 1 of Hands-on Data Analysis with Pandas
# You can calculate this value by following the formula in the textbook or see the example in the solution
# You can also test out the scikit-learn function "MinMaxScaler", which requires transforming the data in 2D array before applying the scaler. This is a common approach in machine learning.
# Your solution

# Min-max scaling, or MinMax normalization, is a feature scaling technique 
# that rescales features to a fixed range, typically, by using the formula: 
# (x - x_min) / (x_max - x_min).

scaled_min_max = [ (x-min(salaries))/ (max(salaries)-min(salaries)) for x in salaries ]
scaled_min_max[:3]

[0.8472361809045226, 0.7608040201005025, 0.4221105527638191]

## Exercise 7(b) Standardizing

In [70]:
# Refer to the Scaling data section of your reading in Chapter 1 of Hands-on Data Analysis with Pandas
# You can calculate this value by following the formula in the textbook or see the example in the solution
# You can also test out the z-score method in Scipy Stats (2D array is needed)
# Your solution

# Z-score normalization, also called standardization, transforms data so that it has a mean (average) of 0 and a standard deviation of 1. 
# This process adjusts data values based on how far they deviate from the mean, measured in units of standard deviation.


mean = np.mean(salaries)
std_dev = np.std(salaries) # By default, np.std uses ddof=0 for population standard deviation

z_scores = [(x - mean) / std_dev for x in salaries]
z_scores[:5]

[np.float64(0.9766171491094446),
 np.float64(0.65146878155336),
 np.float64(-0.6226591238699022),
 np.float64(-1.2351479092662478),
 np.float64(-0.28238757642748796)]

## Chapter 1, Exercise 8

Using the scaled data from exercise 7, calculate the following:

Note. There is some preparation work to create the normalized (scaled) vs. standardized dataset before calculating covariance.

In [71]:
# If you did not setup a scaled data or standardized data with z-score approach, you can use the following codes.
# Scaled data aka normalized data
from sklearn.preprocessing import MinMaxScaler
# Make an array of salaries
salaries_array = np.array(salaries)
# Reshape the data to 2D (one feature)
salaries_2d = salaries_array.reshape(-1, 1)
# Scaled the data
scaler = MinMaxScaler()
scaler.fit(salaries_2d)
# Optionally, transform the data. scaled_salaries in the scaled data for use in next step
scaled_salaries = scaler.transform(salaries_2d)
# Preview the first 10 values. Noticed the double [] that shows the 2D shape
scaled_salaries[:10]
# This will print the shape if you would like to check
# scaled_salaries.shape

array([[0.84723618],
       [0.76080402],
       [0.42211055],
       [0.25929648],
       [0.51256281],
       [0.40603015],
       [0.78693467],
       [0.30351759],
       [0.47839196],
       [0.58492462]])

In [72]:
# Standardized version of the salary data (z-score)
from scipy import stats
# Assign a new variable `standardized_salary` to store the standarized data
standardized_salary = stats.zscore(salaries_array)
# Preview the first 10 values. This is a 1 dimension object
standardized_salary[:10]
# This will print the shape if you would like to check
# standardized_salary.shape

array([ 0.97661715,  0.65146878, -0.62265912, -1.23514791, -0.28238758,
       -0.68315184,  0.74976945, -1.06879293, -0.41093461, -0.01017034])

## Exercise 8(a) The covariance between the standardized and normalized data

In [73]:
# Covariance can be calculated with numpy
# When the (normalized) scaled_salaries varaible was created, it was created as a 2D array.
# We will have to 'flatten' it to match the shape of the standardized_salary before calculation.
scaled_salaries = np.array(scaled_salaries).flatten()

# Now calculate the covariance
cov_calculated = np.cov(scaled_salaries, standardized_salary)
print(cov_calculated)

[[0.07137603 0.26850885]
 [0.26850885 1.01010101]]


Note: How to interpret this output? The covariance function produce a 2X2 covariance matrix, the off-diagonal elements present the covariance of the 2 sets of values. Therefore 0.2685 is the output for intrepretation. Since the covariance estimate is positive, it indicates a positive relationship (moving in the same direction), and we don't know how "strong" this relationship and therefore correlation will help us understand this better.

In [74]:
# Example 2 for covariance calculation
# Covariance can be calculated with statistics module

statistics.covariance(scaled_salaries, standardized_salary)


np.float64(0.2685088459449036)

Note. Getting a correlation of 1 is often very rare, but in this case, we are basically using the same raw data and simply normalized it (force values to 0,1 range) or standardized it (force the data to centered around zero and then deviate from zero) to put it on a common scale. It will not change the strength or direction of the relationship between 2 variables.

# Chapter 2

In [75]:
# The default code to run if you setup your book_env correctly to import the dataset
# df = pd.read_csv('../../ch_02/data/parsed.csv')

# If you encountered issues due to directory setup, you can provide the full file path also
# df = pd.read_csv('cdrive../../ch_02/data/parsed.csv')

## Chapter 2, Exercise 1

Find the 95th percentile of earthquake magnitude in Japan using the mb magnitude type.

In [76]:
# Reading the data by creating a data frame from the parsed.csv
df_parsed = pd.read_csv('./parsed.csv')
df_parsed.head(2)

,alert,cdi,code,detail,dmin,felt,gap,ids,mag,magType,...,status,time,title,tsunami,type,types,tz,updated,url,parsed_place
0,NaN,NaN,37389218,https://earthquake.usgs.gov/fdsnws/event/1/que...,0.008693,NaN,85.0,",ci37389218,",1.35,ml,...,automatic,1539475168010,"M 1.4 - 9km NE of Aguanga, CA",0,earthquake,",geoserve,nearby-cities,origin,phase-data,",-480.0,1539475395144,https://earthquake.usgs.gov/earthquakes/eventp...,California
1,NaN,NaN,37389202,https://earthquake.usgs.gov/fdsnws/event/1/que...,0.020030,NaN,79.0,",ci37389202,",1.29,ml,...,automatic,1539475129610,"M 1.3 - 9km NE of Aguanga, CA",0,earthquake,",geoserve,nearby-cities,origin,phase-data,",-480.0,1539475253925,https://earthquake.usgs.gov/earthquakes/eventp...,California


In [77]:
# Filtering the Japans data with MagType as mb and finding the 95th Percentile on it.
print('95th percentile of earthquake magnitude in Japan is : ',
      df_parsed[ (df_parsed['parsed_place'] == 'Japan') & (df_parsed['magType']=='mb')].mag.quantile(.95))

95th percentile of earthquake magnitude in Japan is :  4.9


## Chapter 2, Exercise 2

Find the percentage of earthquakes in Indonesia that were coupled with tsunamis.

In [78]:
# Checking different values in type, and found there is only one value which is 'earthquake'
df_parsed[ (df_parsed['parsed_place'] == 'Indonesia')].type.unique()

array(['earthquake'], dtype=object)

In [79]:
# Total Earchquake in Indonesia
total_count = len( df_parsed[ (df_parsed['parsed_place'] == 'Indonesia') ].parsed_place )
# Total earthquake in Indonesia with Tsunami presence 
indonesia_tsunami_count = len( df_parsed[ (df_parsed['parsed_place'] == 'Indonesia') & (df_parsed['tsunami']==1)].parsed_place )

perc_indonesia = ( (indonesia_tsunami_count ) / total_count ) *100

print( 'the percentage of earthquakes in Indonesia that were coupled with tsunamis : ' , perc_indonesia)

the percentage of earthquakes in Indonesia that were coupled with tsunamis :  23.12925170068027


## Chapter 2, Exercise 3

Calculate summary statistics for earthquakes in Nevada.

In [80]:
# Checking the Nevada data 

df_parsed[df_parsed.parsed_place=='Nevada'].describe()

,cdi,dmin,felt,gap,mag,mmi,nst,rms,sig,time,tsunami,tz,updated
count,15.000000,681.000000,15.000000,681.000000,681.000000,1.00,681.000000,681.000000,681.000000,6.810000e+02,681.0,681.0,6.810000e+02
mean,2.440000,0.166199,2.400000,153.668120,0.500073,2.84,12.618209,0.151986,10.970631,1.538314e+12,0.0,-480.0,1.538402e+12
std,0.501142,0.166228,4.626013,68.735302,0.696710,NaN,9.866963,0.084662,19.607150,5.965637e+08,0.0,0.0,6.010951e+08
min,2.000000,0.001000,1.000000,29.140000,-0.500000,2.84,3.000000,0.000500,0.000000,1.537247e+12,0.0,-480.0,1.537307e+12
25%,2.000000,0.053000,1.000000,97.380000,-0.100000,2.84,6.000000,0.106900,0.000000,1.537854e+12,0.0,-480.0,1.537928e+12
50%,2.200000,0.112000,1.000000,149.140000,0.400000,2.84,10.000000,0.146300,2.000000,1.538280e+12,0.0,-480.0,1.538428e+12
75%,2.900000,0.233000,1.000000,199.720000,0.900000,2.84,16.000000,0.187100,12.000000,1.538821e+12,0.0,-480.0,1.538878e+12
max,3.300000,1.414000,19.000000,355.910000,2.900000,2.84,61.000000,0.863400,129.000000,1.539461e+12,0.0,-480.0,1.539483e+12


## Chapter 2, Exercise 4

Add a column indicating whether the earthquake happened in a country or US state that is on the Ring of Fire. Use Alaska, Antarctica (look for Antarctic), Bolivia, California, Canada, Chile, Costa Rica, Ecuador, Fiji, Guatemala, Indonesia, Japan, Kermadec Islands, Mexico (be careful not to select New Mexico), New Zealand, Peru, Philippines, Russia, Taiwan, Tonga, and Washington.

In [81]:
# Preparing the list of coutries to pass to the condition 

country_list = ['Alaska', 'Antarctica','Antarctic', 
'Bolivia', 'California', 'Canada', 'Chile', 'Costa Rica', 
'Ecuador', 'Fiji', 'Guatemala', 'Indonesia', 'Japan', 
'Kermadec Islands', 
'Mexico', 
'New Zealand','Peru', 'Philippines', 'Russia', 
'Taiwan', 'Tonga','Washington']
# Creating the new field 'ring_of_fire' using the numpy select function 
df_parsed['ring_of_fire'] = np.select( [df_parsed.parsed_place.isin(country_list) | df_parsed.parsed_place.str.contains('Antarctic') ] ,['Yes'],default='No')
df_parsed.groupby(['ring_of_fire']).parsed_place.count()

ring_of_fire
No     2144
Yes    7188
Name: parsed_place, dtype: int64

## Chapter 2, Exercise 5

Calculate the number of earthquakes in the Ring of Fire locations and the number outside of them.

In [82]:
df_parsed.groupby(['ring_of_fire']).parsed_place.count()

ring_of_fire
No     2144
Yes    7188
Name: parsed_place, dtype: int64

## Chapter 2, Exercise 6

Find the tsunami count along the Ring of Fire.

In [83]:
ring_of_fire_tssunami_count = df_parsed[ (df_parsed.ring_of_fire == 'Yes') & (df_parsed.tsunami ==1) ].parsed_place.count()
print('The tsunami count along the Ring of Fire:', ring_of_fire_tssunami_count)

The tsunami count along the Ring of Fire: 45
